noUnusedLocals checks for type parameters of merged interface should also check merge targets

**TypeScript Version:**  master

**Code**

```ts
interface C<T> {
}
class C<T> {
  public c: T;
}
```

**Expected behavior:**

pass

**Actual behavior:** 

```
$ node built/local/tsc.js --noUnusedLocals index.ts
index.ts(1,13): error TS6133: 'T' is declared but never used.
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -14614,7 +14614,7 @@ namespace ts {
                 if (node.typeParameters) {
                     // Only report errors on the last declaration for the type parameter container;
                     // this ensures that all uses have been accounted for.
-                    const symbol = node.symbol && getMergedSymbol(node.symbol);
+                    const symbol = getSymbolOfNode(node);
                     const lastDeclaration = symbol && symbol.declarations && lastOrUndefined(symbol.declarations);
                     if (lastDeclaration !== node) {
                         return;

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -13274,7 +13274,7 @@ namespace ts {
                         checkAsyncFunctionReturnType(<FunctionLikeDeclaration>node);
                     }
                 }
-                if (!(<FunctionDeclaration>node).body) {
+                if (noUnusedIdentifiers && !(<FunctionDeclaration>node).body) {
                     checkUnusedTypeParameters(node);
                 }
             }
@@ -14609,8 +14609,15 @@ namespace ts {
         function checkUnusedTypeParameters(node: ClassDeclaration | ClassExpression | FunctionDeclaration | MethodDeclaration | FunctionExpression | ArrowFunction | ConstructorDeclaration | SignatureDeclaration | InterfaceDeclaration) {
             if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
                 if (node.typeParameters) {
+                    // Only report errors on the last declaration for the type parameter container;
+                    // this ensures that all uses have been accounted for.
+                    const symbol = node.symbol && getMergedSymbol(node.symbol);
+                    const lastDeclaration = symbol && symbol.declarations && lastOrUndefined(symbol.declarations);
+                    if (lastDeclaration !== node) {
+                        return;
+                    }
                     for (const typeParameter of node.typeParameters) {
-                        if (!typeParameter.symbol.isReferenced) {
+                        if (!getMergedSymbol(typeParameter.symbol).isReferenced) {
                             error(typeParameter.name, Diagnostics._0_is_declared_but_never_used, typeParameter.symbol.name);
                         }
                     }
@@ -16119,7 +16126,7 @@ namespace ts {
 
             if (produceDiagnostics) {
                 checkTypeForDuplicateIndexSignatures(node);
-                checkUnusedTypeParameters(node);
+                registerForUnusedIdentifiersCheck(node);
             }
         }
 

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -2908,7 +2908,6 @@ namespace ts {
         getCancellationToken?(): CancellationToken;
         getDefaultLibFileName(options: CompilerOptions): string;
         getDefaultLibLocation?(): string;
-        getDefaultTypeDirectiveNames?(rootPath: string): string[];
         writeFile: WriteFileCallback;
         getCurrentDirectory(): string;
         getDirectories(path: string): string[];

--- a/tests/baselines/reference/unusedTypeParameters6.js
+++ b/tests/baselines/reference/unusedTypeParameters6.js
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/unusedTypeParameters6.ts] ////
+
+//// [a.ts]
+
+class C<T> { }
+
+//// [b.ts]
+interface C<T> { a: T; }
+
+//// [a.js]
+var C = (function () {
+    function C() {
+    }
+    return C;
+}());
+//// [b.js]

--- a/tests/baselines/reference/unusedTypeParameters6.symbols
+++ b/tests/baselines/reference/unusedTypeParameters6.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/a.ts ===
+
+class C<T> { }
+>C : Symbol(C, Decl(a.ts, 0, 0), Decl(b.ts, 0, 0))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+
+=== tests/cases/compiler/b.ts ===
+interface C<T> { a: T; }
+>C : Symbol(C, Decl(a.ts, 0, 0), Decl(b.ts, 0, 0))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+>a : Symbol(C.a, Decl(b.ts, 0, 16))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+

--- a/tests/baselines/reference/unusedTypeParameters6.types
+++ b/tests/baselines/reference/unusedTypeParameters6.types
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/a.ts ===
+
+class C<T> { }
+>C : C<T>
+>T : T
+
+=== tests/cases/compiler/b.ts ===
+interface C<T> { a: T; }
+>C : C<T>
+>T : T
+>a : T
+>T : T
+

--- a/tests/baselines/reference/unusedTypeParameters7.js
+++ b/tests/baselines/reference/unusedTypeParameters7.js
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/unusedTypeParameters7.ts] ////
+
+//// [a.ts]
+
+class C<T> { a: T; }
+
+//// [b.ts]
+interface C<T> { }
+
+//// [a.js]
+var C = (function () {
+    function C() {
+    }
+    return C;
+}());
+//// [b.js]

--- a/tests/baselines/reference/unusedTypeParameters7.symbols
+++ b/tests/baselines/reference/unusedTypeParameters7.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/a.ts ===
+
+class C<T> { a: T; }
+>C : Symbol(C, Decl(a.ts, 0, 0), Decl(b.ts, 0, 0))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+>a : Symbol(C.a, Decl(a.ts, 1, 12))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+
+=== tests/cases/compiler/b.ts ===
+interface C<T> { }
+>C : Symbol(C, Decl(a.ts, 0, 0), Decl(b.ts, 0, 0))
+>T : Symbol(T, Decl(a.ts, 1, 8), Decl(b.ts, 0, 12))
+

--- a/tests/baselines/reference/unusedTypeParameters7.types
+++ b/tests/baselines/reference/unusedTypeParameters7.types
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/a.ts ===
+
+class C<T> { a: T; }
+>C : C<T>
+>T : T
+>a : T
+>T : T
+
+=== tests/cases/compiler/b.ts ===
+interface C<T> { }
+>C : C<T>
+>T : T
+

--- a/tests/baselines/reference/unusedTypeParameters8.errors.txt
+++ b/tests/baselines/reference/unusedTypeParameters8.errors.txt
@@ -1,0 +1,11 @@
+tests/cases/compiler/b.ts(1,13): error TS6133: 'T' is declared but never used.
+
+
+==== tests/cases/compiler/a.ts (0 errors) ====
+    
+    class C<T> { }
+    
+==== tests/cases/compiler/b.ts (1 errors) ====
+    interface C<T> { }
+                ~
+!!! error TS6133: 'T' is declared but never used.

--- a/tests/baselines/reference/unusedTypeParameters8.js
+++ b/tests/baselines/reference/unusedTypeParameters8.js
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/unusedTypeParameters8.ts] ////
+
+//// [a.ts]
+
+class C<T> { }
+
+//// [b.ts]
+interface C<T> { }
+
+//// [a.js]
+var C = (function () {
+    function C() {
+    }
+    return C;
+}());
+//// [b.js]

--- a/tests/baselines/reference/unusedTypeParameters9.js
+++ b/tests/baselines/reference/unusedTypeParameters9.js
@@ -1,0 +1,30 @@
+//// [unusedTypeParameters9.ts]
+
+// clas + interface
+class C1<T> { }
+interface C1<T> { a: T; }
+
+// interface + class
+class C2<T> { a: T; }
+interface C2<T> { }
+
+// interfaces
+interface C3<T> { a(c: (p: T) => void): void; }
+interface C3<T> { b: string; }
+interface C3<T> { c: number; }
+interface C3<T> { d: boolean;  }
+interface C3<T> { e: any; }
+
+//// [unusedTypeParameters9.js]
+// clas + interface
+var C1 = (function () {
+    function C1() {
+    }
+    return C1;
+}());
+// interface + class
+var C2 = (function () {
+    function C2() {
+    }
+    return C2;
+}());

--- a/tests/baselines/reference/unusedTypeParameters9.symbols
+++ b/tests/baselines/reference/unusedTypeParameters9.symbols
@@ -1,0 +1,53 @@
+=== tests/cases/compiler/unusedTypeParameters9.ts ===
+
+// clas + interface
+class C1<T> { }
+>C1 : Symbol(C1, Decl(unusedTypeParameters9.ts, 0, 0), Decl(unusedTypeParameters9.ts, 2, 15))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 2, 9), Decl(unusedTypeParameters9.ts, 3, 13))
+
+interface C1<T> { a: T; }
+>C1 : Symbol(C1, Decl(unusedTypeParameters9.ts, 0, 0), Decl(unusedTypeParameters9.ts, 2, 15))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 2, 9), Decl(unusedTypeParameters9.ts, 3, 13))
+>a : Symbol(C1.a, Decl(unusedTypeParameters9.ts, 3, 17))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 2, 9), Decl(unusedTypeParameters9.ts, 3, 13))
+
+// interface + class
+class C2<T> { a: T; }
+>C2 : Symbol(C2, Decl(unusedTypeParameters9.ts, 3, 25), Decl(unusedTypeParameters9.ts, 6, 21))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 6, 9), Decl(unusedTypeParameters9.ts, 7, 13))
+>a : Symbol(C2.a, Decl(unusedTypeParameters9.ts, 6, 13))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 6, 9), Decl(unusedTypeParameters9.ts, 7, 13))
+
+interface C2<T> { }
+>C2 : Symbol(C2, Decl(unusedTypeParameters9.ts, 3, 25), Decl(unusedTypeParameters9.ts, 6, 21))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 6, 9), Decl(unusedTypeParameters9.ts, 7, 13))
+
+// interfaces
+interface C3<T> { a(c: (p: T) => void): void; }
+>C3 : Symbol(C3, Decl(unusedTypeParameters9.ts, 7, 19), Decl(unusedTypeParameters9.ts, 10, 47), Decl(unusedTypeParameters9.ts, 11, 30), Decl(unusedTypeParameters9.ts, 12, 30), Decl(unusedTypeParameters9.ts, 13, 32))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+>a : Symbol(C3.a, Decl(unusedTypeParameters9.ts, 10, 17))
+>c : Symbol(c, Decl(unusedTypeParameters9.ts, 10, 20))
+>p : Symbol(p, Decl(unusedTypeParameters9.ts, 10, 24))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+
+interface C3<T> { b: string; }
+>C3 : Symbol(C3, Decl(unusedTypeParameters9.ts, 7, 19), Decl(unusedTypeParameters9.ts, 10, 47), Decl(unusedTypeParameters9.ts, 11, 30), Decl(unusedTypeParameters9.ts, 12, 30), Decl(unusedTypeParameters9.ts, 13, 32))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+>b : Symbol(C3.b, Decl(unusedTypeParameters9.ts, 11, 17))
+
+interface C3<T> { c: number; }
+>C3 : Symbol(C3, Decl(unusedTypeParameters9.ts, 7, 19), Decl(unusedTypeParameters9.ts, 10, 47), Decl(unusedTypeParameters9.ts, 11, 30), Decl(unusedTypeParameters9.ts, 12, 30), Decl(unusedTypeParameters9.ts, 13, 32))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+>c : Symbol(C3.c, Decl(unusedTypeParameters9.ts, 12, 17))
+
+interface C3<T> { d: boolean;  }
+>C3 : Symbol(C3, Decl(unusedTypeParameters9.ts, 7, 19), Decl(unusedTypeParameters9.ts, 10, 47), Decl(unusedTypeParameters9.ts, 11, 30), Decl(unusedTypeParameters9.ts, 12, 30), Decl(unusedTypeParameters9.ts, 13, 32))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+>d : Symbol(C3.d, Decl(unusedTypeParameters9.ts, 13, 17))
+
+interface C3<T> { e: any; }
+>C3 : Symbol(C3, Decl(unusedTypeParameters9.ts, 7, 19), Decl(unusedTypeParameters9.ts, 10, 47), Decl(unusedTypeParameters9.ts, 11, 30), Decl(unusedTypeParameters9.ts, 12, 30), Decl(unusedTypeParameters9.ts, 13, 32))
+>T : Symbol(T, Decl(unusedTypeParameters9.ts, 10, 13), Decl(unusedTypeParameters9.ts, 11, 13), Decl(unusedTypeParameters9.ts, 12, 13), Decl(unusedTypeParameters9.ts, 13, 13), Decl(unusedTypeParameters9.ts, 14, 13))
+>e : Symbol(C3.e, Decl(unusedTypeParameters9.ts, 14, 17))
+

--- a/tests/baselines/reference/unusedTypeParameters9.types
+++ b/tests/baselines/reference/unusedTypeParameters9.types
@@ -1,0 +1,53 @@
+=== tests/cases/compiler/unusedTypeParameters9.ts ===
+
+// clas + interface
+class C1<T> { }
+>C1 : C1<T>
+>T : T
+
+interface C1<T> { a: T; }
+>C1 : C1<T>
+>T : T
+>a : T
+>T : T
+
+// interface + class
+class C2<T> { a: T; }
+>C2 : C2<T>
+>T : T
+>a : T
+>T : T
+
+interface C2<T> { }
+>C2 : C2<T>
+>T : T
+
+// interfaces
+interface C3<T> { a(c: (p: T) => void): void; }
+>C3 : C3<T>
+>T : T
+>a : (c: (p: T) => void) => void
+>c : (p: T) => void
+>p : T
+>T : T
+
+interface C3<T> { b: string; }
+>C3 : C3<T>
+>T : T
+>b : string
+
+interface C3<T> { c: number; }
+>C3 : C3<T>
+>T : T
+>c : number
+
+interface C3<T> { d: boolean;  }
+>C3 : C3<T>
+>T : T
+>d : boolean
+
+interface C3<T> { e: any; }
+>C3 : C3<T>
+>T : T
+>e : any
+

--- a/tests/cases/compiler/unusedTypeParameters6.ts
+++ b/tests/cases/compiler/unusedTypeParameters6.ts
@@ -1,0 +1,8 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
+
+// @fileName: a.ts
+class C<T> { }
+
+// @fileName: b.ts
+interface C<T> { a: T; }

--- a/tests/cases/compiler/unusedTypeParameters7.ts
+++ b/tests/cases/compiler/unusedTypeParameters7.ts
@@ -1,0 +1,8 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
+
+// @fileName: a.ts
+class C<T> { a: T; }
+
+// @fileName: b.ts
+interface C<T> { }

--- a/tests/cases/compiler/unusedTypeParameters8.ts
+++ b/tests/cases/compiler/unusedTypeParameters8.ts
@@ -1,0 +1,8 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
+
+// @fileName: a.ts
+class C<T> { }
+
+// @fileName: b.ts
+interface C<T> { }

--- a/tests/cases/compiler/unusedTypeParameters9.ts
+++ b/tests/cases/compiler/unusedTypeParameters9.ts
@@ -1,0 +1,17 @@
+//@noUnusedLocals:true
+//@noUnusedParameters:true
+
+// clas + interface
+class C1<T> { }
+interface C1<T> { a: T; }
+
+// interface + class
+class C2<T> { a: T; }
+interface C2<T> { }
+
+// interfaces
+interface C3<T> { a(c: (p: T) => void): void; }
+interface C3<T> { b: string; }
+interface C3<T> { c: number; }
+interface C3<T> { d: boolean;  }
+interface C3<T> { e: any; }

